lgc/patch/PatchReadFirstLane.cpp

```diff
@@ -125,21 +106,61 @@ static bool areAllUserReadFirstLane(Instruction *inst) {
   return true;
 }
 
-PatchReadFirstLane::PatchReadFirstLane() : FunctionPass(ID), m_targetTransformInfo(nullptr) {
+// =====================================================================================================================
+PatchReadFirstLane::PatchReadFirstLane() : m_targetTransformInfo(nullptr) {
+}
+
+// =====================================================================================================================
+LegacyPatchReadFirstLane::LegacyPatchReadFirstLane() : FunctionPass(ID) {
+}
+
+// =====================================================================================================================
+// Executes this LLVM pass on the specified LLVM function.
+//
+// @param [in/out] function : Function that we will peephole optimize.
+// @returns : True if the module was modified by the transformation and false otherwise
+bool LegacyPatchReadFirstLane::runOnFunction(Function &function) {
+  auto *targetTransformInfoWrapperPass = getAnalysisIfAvailable<TargetTransformInfoWrapperPass>();
+  TargetTransformInfo *targetTransformInfo;
+  if (targetTransformInfoWrapperPass)
+    targetTransformInfo = &targetTransformInfoWrapperPass->getTTI(function);
+  else
+    llvm_unreachable("TTI should be available");
+
+  LegacyDivergenceAnalysis *divergenceAnalysis = &getAnalysis<LegacyDivergenceAnalysis>();
+  auto isDivergentUse = [&](const Use &use) { return divergenceAnalysis->isDivergentUse(&use); };
+  return m_impl.runImpl(function, isDivergentUse, targetTransformInfo);
+}
+
+// =====================================================================================================================
+// Executes this LLVM pass on the specified LLVM function.
+//
+// @param [in/out] function : Function that we will peephole optimize.
+// @param [in/out] analysisManager : Analysis manager to use for this transformation
+// @returns : The preserved analyses (The Analyses that are still valid after this pass)
+PreservedAnalyses PatchReadFirstLane::run(Function &function, FunctionAnalysisManager &analysisManager) {
+  TargetTransformInfo &targetTransformInfo = analysisManager.getResult<TargetIRAnalysis>(function);
+
+  DivergenceInfo &divergenceInfo = analysisManager.getResult<DivergenceAnalysis>(function);
+  auto isDivergentUse = [&](const Use &use) { return divergenceInfo.isDivergentUse(use); };
```

Comment:
This is a re-definition of the code inside `runOnFunction`. Maybe you can extract this into a sub-method that creates the lambda or a internal class which overrides `()`.

Reply:
The code is actually different from what is inside `runOnFunction`. One is for the LLVM DivergenceAnalysis pass, the other is for the LLVM LegacyDivergenceAnalysis pass.